db/snapshot/backup_task.cc

```diff
@@ -99,52 +111,132 @@ future<> backup_task_impl::upload_component(sstring name) {
+    co_return greedy_partitioning(std::move(backup_files), smp::count);
+}
+} // namespace
 
-        std::ignore = upload_component(component_ent->name).handle_exception([&ex] (std::exception_ptr e) {
-            // keep the first exception
-            if (!ex) {
-                ex = std::move(e);
-            }
-        }).finally([gh = std::move(gh)] {});
-        co_await coroutine::maybe_yield();
-        co_await utils::get_local_injector().inject("backup_task_pause", utils::wait_for_message(std::chrono::minutes(2)));
-        if (impl::_as.abort_requested()) {
-            ex = impl::_as.abort_requested_exception_ptr();
-            break;
-        }
+future<> backup_task_impl::do_backup() {
+    if (!co_await file_exists(_snapshot_dir.native())) {
+        throw std::invalid_argument(fmt::format("snapshot does not exist at {}", _snapshot_dir.native()));
     }
 
-    co_await snapshot_dir_lister.close();
-    co_await uploads.close();
-    if (ex) {
+    auto backup_shards = co_await get_backup_files(_snapshot_dir);
+
+    if (snap_log.is_enabled(log_level::debug))
+        snap_log.debug("{}", print_results(backup_shards));
+
+    std::vector<abort_source> as_per_shard(smp::count);
+    auto s = _as.subscribe([&]() noexcept {
+        try {
+            std::ignore = smp::invoke_on_all([&as_per_shard, ex = _as.abort_requested_exception_ptr()] {
+                as_per_shard[this_shard_id()].request_abort_ex(ex);
+            });
+        } catch (...) {
+        }
+    });
+
+    auto [_1, memory, _2] = _cfg();
+    auto shard_mem = memory.current() / smp::count;
+    std::exception_ptr ex;
+    co_await smp::invoke_on_all([this, &ex, &backup_shards, &shard_mem, &as_per_shard] -> future<> {
+        auto shard_id = this_shard_id();
+        const auto& shard_files = std::get<1>(backup_shards[shard_id]);
+        auto shard_semaphore = semaphore(shard_mem);
+        auto& shard_as = as_per_shard[shard_id];
+        auto& shard_progress = _progress_per_shard[shard_id];
+
+        auto [endpoint, _1, cfg] = _cfg();
+        auto cln = s3::client::make(endpoint, make_lw_shared<s3::endpoint_config>(std::move(cfg)), shard_semaphore);
+
+        gate uploads;
+
+        for (const auto& name : shard_files) {
+            // Pre-upload break point. For testing abort in actual s3 client usage.
+            co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+            auto gh = uploads.hold();
+
+            std::ignore = upload_component(cln, shard_as, shard_progress, name)
+                              .handle_exception([&ex](std::exception_ptr e) {
+                                  if (!ex) {
+                                      ex = std::move(e);
+                                  }
+                              })
+                              .finally([gh = std::move(gh)] {});
```

Comment:
Please don't mix continuation style and coroutines. Also add a comment explaining why ignoring the future is safe.

Reply:
it is existing code but I will gladly coroutinize it